HPCC-15142 Minimal changes needed for 'make install'

Michael-Gardner opened the pull request:
Workflow:
export DESTDIR=$HOME/testdir
cmake -DRUNTIME_USER=$USER ../HPCC-Platform/
make install
make configure
$DESTDIR/etc/init.d/hpcc-init start

Signed-off-by: Michael Gardner michael.gardner@lexisnexis.com

Michael-Gardner committed:
HPCC-15142 Minimal changes needed for DESTDIR

Workflow:
export DESTDIR=$HOME/testdir
cmake -DRUNTIME_USER=$USER ../HPCC-Platform/
make install
make configure
$DESTDIR/etc/init.d/hpcc-init start

Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

Michael-Gardner commented:
@ghalliday @xwang2713 can you please review?  I also have a matching LN PR going up that has a very minor change to make sure componentfiles get installed to the correct destination.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 162ce7e31d7c9761ad21e43ab6dbb9088672ef5c
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 86388204b1f521a48f5ae950e9e026495365c8f4
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on initfiles/etc/DIR_NAME/CMakeLists.txt:
```diff
@@ -15,17 +15,26 @@
-Install ( FILES ${CMAKE_CURRENT_BINARY_DIR}/environment.xml DESTINATION .${CONFIG_DIR}/rpmnew COMPONENT Runtime ) 	#  Don't delete the "."
-Install ( FILES ${CMAKE_CURRENT_BINARY_DIR}/environment.conf DESTINATION .${CONFIG_DIR}/rpmnew COMPONENT Runtime )
-Install ( FILES ${CMAKE_CURRENT_SOURCE_DIR}/genenvrules.conf DESTINATION .${CONFIG_DIR}/rpmnew COMPONENT Runtime )
+Install ( FILES ${CMAKE_CURRENT_BINARY_DIR}/environment.xml DESTINATION etc/${DIR_NAME}/rpmnew COMPONENT Runtime ) 	#  Don't delete the "."
```
trivial: there is no longer any . to not delete!

ghalliday commented:
I had a very quick scan through and the changes looked sensible.  Where will it be documented how to use this new functionality?
There should be comments in the cmake file describing what options you can set and what effect it will have.  Not sure where else.

Michael-Gardner committed:
HPCC-15142 Documentation Changes for make install

* Root level CMakeLists.txt documentation changes for building
   a development installation
* Modifications to initfiles/etc/DIR_NAME/CMakeLists.txt for
   cleanup of unnecessary comments

Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

Michael-Gardner commented:
@ghalliday I've gone ahead and added the documentation necessary to build a development installation of the platform.  I put it in the root level CMakeLists.txt file, with the directions to build a normal windows or Linux platform.

@g-pan Should we be documenting how to build a development installation anywhere other than the cmake files?

HPCCSmoketest commented:
Automated Smoketest 
Sha: 8eb7a89b8505062f5542a5109a6e39824a909dcc
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

g-pan commented:
In HPCCPlatform git wiki https://github.com/hpcc-systems/HPCC-Platform/wiki/Building-HPCC

Michael-Gardner commented:
@ghalliday https://github.com/hpcc-systems/HPCC-Platform/wiki/Building-HPCC#to-create-a-development-installation-on-linux Documentation added to the wiki

xwang2713 commented:
Looks good. User still need run with sudo if not root, right?

Michael-Gardner commented:
No.  You don't have to raise permissions at all.

Michael-Gardner commented:
@richardkchapman ready for you to look at.

richardkchapman commented:
Why is DESTDIR passed in to the cmake process via an environment variable rather than using -DDESTDIR=xxx ?

Michael-Gardner commented:
Under the hood, cmake is using $ENV{DESTDIR} in the cmake installation scripts that get generated.  Which means if you pass it in with -DDESTDIR but don't have it set as an environment variable, it won't prefix itself to $CMAKE_INSTALL_PREFIX for free.  I could use -DDESTDIR=xxx but it would just be more work for something that cmake already does on it's own.

richardkchapman commented:
The problem with needing it set via an environment variable is that it won't be preserved in the CMakeCache.txt - so if I set DESTDIR when I run cmake, but then run make which causes cmake to be rerun it's liable to lose the setting and confusion will about. Also there's no logical reason from the user's point of view that this one value is set via a different mechanism to all the others... just because it "came for free" for the developer

Michael-Gardner commented:
Ok.  I'll change to to take the -DDESTDIR instead

richardkchapman commented:
Testing this on OSX (where, if it works, it will make my life MUCH easier!)

Get lots of messages of the form

error: /Library/Developer/CommandLineTools/usr/bin/install_name_tool: input file: /Users/rchapman/var/opt/HPCCSystems/bin/run_keydiff is not a Mach-O file

I don't think they are fatal but would be nice to suppress them if possible (I suspect they have always been present but hidden away in logs that are not usually visible).

richardkchapman commented:
)$ ~/var/etc/init.d/hpcc-init start
chown: rchapman: illegal group name
chown: rchapman: illegal group name
date: illegal option -- -
usage: date [-jnu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ... 
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
date: illegal option -- -
usage: date [-jnu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ... 
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
stat: illegal option -- c
usage: stat [-FlLnqrsx] [-f format] [-t timefmt] [file ...]
/Users/rchapman/var/opt/HPCCSystems/etc/init.d/hpcc_common: line 743: [: =: unary operator expected
date: illegal option -- -
usage: date [-jnu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ... 
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
Group for user rchapman does not exist on the system. Exiting ...

Michael-Gardner commented:
Try setting -DRUNTIME_GROUP=valid group for your rchapman user.  That will override the default setting where the runtime group matches your runtime user.

richardkchapman commented:
I set -DRUNTIME_GROUP=staff

And now I get:

chown: rchapman: illegal group name
chown: rchapman: illegal group name
date: illegal option -- -
usage: date [-jnu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ... 
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
date: illegal option -- -
usage: date [-jnu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ... 
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
stat: illegal option -- c
usage: stat [-FlLnqrsx] [-f format] [-t timefmt] [file ...]
/Users/rchapman/var/opt/HPCCSystems/etc/init.d/hpcc_common: line 743: [: =: unary operator expected
date: illegal option -- -
usage: date [-jnu] [-d dst] [-r seconds] [-t west] [-v[+|-]val[ymwdHMS]] ... 
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
Group for user staff does not exist on the system. Exiting ...

richardkchapman commented:
A search for chown in the sources finds a few places that look like they are wrong:

e.g. hpcc-init.in
install-init.in

Michael-Gardner commented:
Richard, the machine we have in the office for osx is a little outdated.  I'm going to work on getting it set up for development, but for now here are some changes to date/chown/stat that are necessary for gnu/freebsd portability.

HPCCSmoketest commented:
Automated Smoketest 
Sha: ad7ee4bd84dcb326188e0862848b3524aca3c6a8
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
See LN PR for details of a build failure with LN system.

richardkchapman commented:
Got a bit further - now I see:

$ ~/var/etc/init.d/hpcc-init start
Exception: 21:
CFile::open /Users/rchapman/var/etc/HPCCSystems/, Is a directory
There are no components configured to run on this node ...

HPCCSmoketest commented:
Automated Smoketest 
Sha: c5b94b8b05dab756395e3c68e39189c55cdb3403
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

Michael-Gardner commented:
Updated PR.  Changed DESTDIR from an environmental variable to a cmake variable.  Now instead of using an 'export DESTDIR' before you run the cmake command, just add '-DDESTDIR=$HOME/myruntime'.  @ghalliday @richardkchapman The reason there was a CFile::open issue was because there isn't a configuration file automatically written to /etc/HPCCSystems/environment.xml.  run 'make configure' if you're using the DESTDIR option to populate the environment.xml and conf files.

Michael-Gardner commented:
The bsd version of sed that is used in osx is stripping the 't' character from environment=environment.xml.  sed 's/[ \t]*=[ \t]*/=/'   The \t is being seen as a 't' character instead of a tab character.  I changed it over to [:blank:] for posix compliance and it seems to work in my tests.

Looks like there is going to be an issue with getopt as well.

ghalliday commented:
Any ideas why I get the following:

gavin:~/builddev$ ~/hpcc/etc/init.d/hpcc-init start
/home/gavin/hpcc/opt/HPCCSystems/etc/init.d/hpcc_common: line 266: cfg.section.DEFAULT: command not found
/home/gavin/hpcc/etc/init.d/hpcc-init: line 126: /sbin/hpcc_setenv: No such file or directory
Group for user  does not exist on the system. Exiting ...

ghalliday commented:
My problem was almost certainly not reading the instructions - not doing a make configure...  I'll try again.

Michael-Gardner commented:
@richardkchapman I have getopts working fine now for osx.  There will be a slight change in the hpcc-init and dafilesrv code, but if you do a 'brew install gnu-getopts && brew link --force gnu-getopts' it will work.  I'm working on another commit with these changes necessary for osx.  Should have it up after lunch.

Michael-Gardner committed:
HPCC-15142 DESTDIR moved to cmake variable

Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

Michael-Gardner committed:
HPCC-15142 OSX compatibility changes

Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

Michael-Gardner commented:
Modified commit history so all changes pertaining to OSX compatibility are in the same commit.

HPCCSmoketest commented:
Automated Smoketest 
Sha: c9ba63c969248f5f9785207871ab227b0d6d5001
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 67a6d95edd26bffa41a920c0984d699a017cae9f
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

Michael-Gardner commented:
@richardkchapman I have it working almost with osx.  Only issue I'm running into is loading plugins into esp.

00000008 2016-08-09 13:12:20.143 96457    -1 "ERROR: ESP Failed to load shared object (liblibws_fs.so.dylib)"
00000009 2016-08-09 13:12:20.143 96457    -1 "ERROR: ESP Unhandled IException (-1 -- procedure esp_service_factory can't be loaded)"


No idea why it's trying to load liblibws_fs.so.dylib with that name.  It's on the system correctly with libws_fs.dylib.

cmake command I'm using to generate the makefiles is:
 cmake -DUSE_OPENLDAP=OFF -DDESTDIR=$HOME/runtime -DRUNTIME_USER=$USER -DRUNTIME_GROUP=staff -DCMAKE_BUILD_TYPE=Release ../HPCC-Platform/

Michael-Gardner commented:
The esp isn't starting because several esp services and bindings are generated in the xml with lib and .so added to the filenames.  It's an issue with componentfiles/configxml.  Do you want me to remove the last commit and make a new Jira for osx compatibility for 6.2.0 or do it all here?

richardkchapman commented:
@Michael-Gardner Do it all here is fine

Michael-Gardner commented:
@garonsky can you please look at the xsl changes in the OSX compatibility commit.  (https://github.com/hpcc-systems/HPCC-Platform/pull/8956/commits/007b784d062a859181e2aacf50237f7111251649)  The added lib < plugin file name > .so isn't really needed because the code will take the library file and make sure it has the correct prefix/suffix so long as it has a base name in the generated xml (ex ws_fs turns into libws_fs.so, or libws_fs.dylib when it goes to load the plugin.)

@richardkchapman Ready for you to take a look at it again on OSX.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 007b784d062a859181e2aacf50237f7111251649
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:82 passed:82 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

garonsky commented:
xsl changes look good.

ghalliday commented:
I tried again, including the make configure and it failed to run.  Nothing was output for a while, and then text like the following:

gavin:~/builddev$ ~/hpcc/etc/init.d/hpcc-init start
rm: cannot remove ‘/var/log/HPCCSystems/hpcc-init.debug’: Permission denied
touch: cannot touch ‘/var/log/HPCCSystems/hpcc-init.debug’: Permission denied
chown: changing ownership of ‘/var/log/HPCCSystems/hpcc-init.debug’: Operation not permitted
/home/gavin/hpcc/etc/init.d/hpcc-init: line 287: /var/log/HPCCSystems/hpcc-init.debug: Permission denied
+ '[' -z ']'
+ (( i=0 ))
+ (( i<=8 ))
+ component=' mydali'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent myeclccserver'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent myeclccserver myeclscheduler'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent myeclccserver myeclscheduler myesp'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent myeclccserver myeclscheduler myesp myroxie'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent myeclccserver myeclscheduler myesp myroxie mysasha'
+ (( i++  ))
+ (( i<=8 ))
+ component=' mydali mydfuserver myeclagent myeclccserver myeclscheduler myesp myroxie mysasha mythor'


The tracing in hpcc-init.log contains:

2016_08_09_09_24_56: --------------------------
2016_08_09_09_24_56: --------------------------
2016_08_09_09_24_56: Group for user  does not exist on the system. Exiting ...
2016_08_09_09_39_21: --------------------------
2016_08_09_09_39_21: --------------------------
2016_08_09_09_39_21: Group for user  does not exist on the system. Exiting ...
2016_08_09_09_39_50: --------------------------
2016_08_09_09_39_50: --------------------------
2016_08_09_09_39_50: Group for user  does not exist on the system. Exiting ...
2016_08_11_10_50_47: --------------------------
2016_08_11_10_50_47: --------------------------

My CMakeCache.txt contains

RUNTIME_USER:BOOL=gavin

(Is the :BOOL significant?)

richardkchapman commented:
One gotcha - not sure if there's much we can do, but when I used -DDESTDIR=~/var I got errors as the ~ was never expanded out to $HOME. Setting -DDESTDIR=/Users/rchapman/var got further...

richardkchapman commented:
Now I am getting

$ ~/var/etc/init.d/hpcc-init start
Usage: /Users/rchapman/var/etc/init.d/hpcc-init [-c component] {start|stop|restart|status|setup}
       /Users/rchapman/var/etc/init.d/hpcc-init [--componentlist] Display node component name list.
       /Users/rchapman/var/etc/init.d/hpcc-init [--typelist] Display node component type list.
       /Users/rchapman/var/etc/init.d/hpcc-init [-f] [--force] Force hard kill on orphaned process groups
       /Users/rchapman/var/etc/init.d/hpcc-init [-h] [--help] Display help

Components on this node as defined by /etc/HPCCSystems/environment.xml:
mydali
mydfuserver
myeclagent
myeclccserver
myeclscheduler
myesp
myroxie
mysasha
mythor

ghalliday commented:
@Michael-Gardner

Michael-Gardner commented:
@ghalliday The debugging info being piped out to your terminal is because we exec stderr to a file, but the fd has a permissions issue in your case, so it fails and the -x flag dumps stderr to your terminal.

The  rm: cannot remove ‘/var/log/HPCCSystems/hpcc-init.debug’: Permission denied is key here.  You're trying to touch a file under root, which of course you aren't going to have permissions for.  This generally is because -DDESTDIR= wasn't set.

What I suspect happened was you were using a build directory that previously had makefiles built without -DDESTDIR=, or with DESTDIR as an environment variable, and the DESTDIR didn't get updated in the configured scripts properly when it changed.

1. Clean out your build directory completely w/ an rm -rf.  Lets clean out the cmake cache to make sure everything gets set properly.
2. cmake -DDESTDIR=/home/gavin/runtime -DRUNTIME_USER=gavin ../HPCC-Platform/
3. make -j install

@richardkchapman (mac only) Your issue is probably related to getopt.  Did you do 'brew install gnu-getopt && brew link --force gnu-getopt' ?  And if you do 'which getopt' does it show the path as /usr/local/bin/getopt?  The getopt that is supplied with freebsd utilities doesn't allow for long options.  So when we pass our options to getopt it just spits them back out at us without doing anything.  Which means when we go to parse the return we get the -- case and print_usage is ran.

richardkchapman commented:
Updating getopt as you describe means I get further (now I get the same issue as Gavin...)

richardkchapman commented:
(if we could detect the old version of getopt and give a helpful error message it would save grief)

richardkchapman commented:
Cleaning everything then doing a make install has given me a new error:

CMake Error at esp/src/cmake_install.cmake:47 (message):
  ERROR: node.js is required to build - see
  https://nodejs.org/en/download/package-manager/

Michael-Gardner commented:
brew install nodejs should be all you need.

richardkchapman commented:
Getting closer

Anything that tries to load the java plugin is failing - a version mismatch I think (I will investigate and/or disable the java plugin).

Thor won't start as it is prompting for a password:

$ ~/var/etc/init.d/hpcc-init start
Starting mydafilesrv ...       [   OK    ]   
Starting mydali ...            [   OK    ]   
Starting mydfuserver ...       [   OK    ]   
Starting myeclagent ...        [   OK    ]   
Starting myeclccserver ...     [   OK    ]   
Starting myeclscheduler ...    [   OK    ]   
Starting myesp ...             [ FAILED  ]   
Starting myroxie ...           [ FAILED  ]   
Starting mysasha ...           [   OK    ]   
Starting mythor ...           Password:
Password:
Password:

richardkchapman commented:
Disabled the eclwatch plugin and ESP starts, but it's giving me error 404 on all the css and js files and giving a blank page as a result.

richardkchapman commented:
I think the eclwatch issues were because I had a duff cached build of the eclwatch js files - I forced a rebuild and that is working now. Just java plugin and thor issues

richardkchapman commented:
Successfully ran queries from ECL playground against hThor and Roxie

richardkchapman commented:
I think this is ok to merge, but @Michael-Gardner can you raise JIRAs for some of the improvements in error messages highlighted in the above conversation
